pkg/cmd/experimental/buildchain/buildchain.go

```diff
@@ -88,215 +90,101 @@ func NewEdge(fullname, to string) *Edge {
 }
 
 // NewCmdBuildChain implements the OpenShift experimental build-chain command
-func NewCmdBuildChain(f *clientcmd.Factory, parentName, name string) *cobra.Command {
+func NewCmdBuildChain(fullName string, f *clientcmd.Factory, out io.Writer) *cobra.Command {
+	options := &BuildChainOptions{}
 	cmd := &cobra.Command{
-		Use:     fmt.Sprintf("%s [IMAGESTREAM:TAG | --all]", name),
+		Use:     "build-chain [IMAGESTREAM:TAG | --all]",
 		Short:   "Output build dependencies of a specific image stream",
 		Long:    buildChainLong,
-		Example: buildChainExample,
+		Example: fmt.Sprintf(buildChainExample, fullName),
 		Run: func(cmd *cobra.Command, args []string) {
-			err := RunBuildChain(f, cmd, args)
-			cmdutil.CheckErr(err)
+			if err := options.Complete(f, cmd, args, out); err != nil {
+				cmdutil.CheckErr(err)
+			}
+			if err := options.Validate(cmd, args); err != nil {
+				cmdutil.CheckErr(cmdutil.UsageError(cmd, err.Error()))
+			}
+			if err := options.RunBuildChain(f, cmd, args); err != nil {
+				cmdutil.CheckErr(err)
+			}
 		},
 	}
 
-	cmd.Flags().Bool("all", false, "Build dependency trees for all image streams")
-	cmd.Flags().Bool("all-tags", false, "Build dependency trees for all tags of a specific image stream")
-	cmd.Flags().StringP("output", "o", "json", "Output format of dependency tree(s)")
+	cmd.Flags().BoolVar(&options.allNamespaces, "all", false, "Build dependency trees for all image streams")
+	cmd.Flags().BoolVar(&options.allTags, "all-tags", false, "Build dependency trees for all tags of a specific image stream")
+	cmd.Flags().StringVarP(&options.output, "output", "o", "json", "Output format of dependency tree(s)")
 	return cmd
 }
 
-// RunBuildChain contains all the necessary functionality for the OpenShift
-// experimental build-chain command
-func RunBuildChain(f *clientcmd.Factory, cmd *cobra.Command, args []string) error {
-	all := cmdutil.GetFlagBool(cmd, "all")
-	allTags := cmdutil.GetFlagBool(cmd, "all-tags")
-	if len(args) > 1 ||
-		(len(args) == 1 && all) ||
-		(len(args) == 0 && allTags) ||
-		(all && allTags) {
-		return cmdutil.UsageError(cmd, "Must pass nothing, an ImageStream name:tag combination, or specify the --all flag")
-	}
-
+// Complete completes the required options for build-chain
+func (o *BuildChainOptions) Complete(f *clientcmd.Factory, cmd *cobra.Command, args []string, out io.Writer) error {
+	// Setup client
 	oc, _, err := f.Clients()
 	if err != nil {
 		return err
 	}
+	o.oc = oc
 
-	// Retrieve namespace(s)
-	namespace := cmdutil.GetFlagString(cmd, "namespace")
-	if len(namespace) == 0 {
-		namespace, _, err = f.DefaultNamespace()
-		if err != nil {
-			return err
-		}
-	}
-	namespaces := make([]string, 0)
-	if all {
+	// Setup namespace
+	if o.allNamespaces {
```

Comment:
@bparees Should `--all` mean "all namespaces you can see" or "all namespaces".  The former is more useful for users, the latter is more useful for admins.  I'd be inclined to have `oc build-chain` do all that you can see and `oadm build-chain` use `kapi.NamespaceAll`